src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,892 @@
+# NOTE: the fields ring_iso and mat_iso are always defined if the field X is
+"""
+    MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+Type of groups `G` of `n x n` matrices over the ring `R`, where `n = degree(G)` and `R = base_ring(G)`.
+
+At the moment, only rings of type `FqNmodFiniteField` are supported.
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+   deg::Int
+   ring::Ring
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   descr::Symbol                       # e.g. GL, SL, symbols for isometry groups
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+   order::fmpz
+   AbstractAlgebra.@declare_other
+
+   MatrixGroup(m::Int, F::Ring) = new{elem_type(F), dense_matrix_type(elem_type(F))}(m,F)
+
+end
+
+# build a MatrixGroup given a list of generators, given as array of either MatrixGroupElem or AbstractAlgebra matrices
+# WARNING: if the elements of V have type MatElem, it does not check whether the determinant is nonzero
+function MatrixGroup(m::Int, F::Ring, V::AbstractVector{T}) where T<:Union{MatElem,AbstractMatrixGroupElem}
+   G = MatrixGroup(m,F)
+
+   L = Vector{MatrixGroupElem}(undef, length(V))
+   for i in 1:length(V)
+      @assert base_ring(V[i])==F "The elements must have the same base ring"
+      @assert nrows(V[i])==m "The elements must have the same dimension"
+      if T<:MatElem
+         L[i] = MatrixGroupElem(G,V[i])
+      else
+# TODO: this part of code can be replaced by the following line
+# L[i] = deepcopy(V[i]); L[i].parent = G;
+         if isdefined(V[i],:elm)
+            if isdefined(V[i],:X)
+               L[i] = MatrixGroupElem(G,V[i].elm,V[i].X)
+            else
+               L[i] = MatrixGroupElem(G,V[i].elm)
+            end
+         else
+            L[i] = MatrixGroupElem(G,V[i].X)
+         end
+# once deepcopy works for GAP objects
+      end
+   end
+   G.gens = L
+
+   return G
+end
+
+
+# NOTE: at least one of the fields :elm and :X must always defined, but not necessarily both of them.
+"""
+    MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: GAPGroupElem{MatrixGroup}
+
+Elements of a group of type `MatrixGroup{RE<:RingElem, T<:MatElem{RE}}`
+"""
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T                         # Oscar matrix
+   X::GapObj                     # GAP matrix. If x isa MatrixGroupElem, then x.X = x.parent.mat_iso(x.elm)
+
+   #MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x::MatElem{RE}) where RE <: RingElem = new{RE, MatElem{RE}}(G,x)
+   MatrixGroupElem(G::MatrixGroup, x::MatElem) = new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}(G,x)
+
+   #function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x_gap::GapObj) where RE <: RingElem
+   function MatrixGroupElem(G::MatrixGroup, x_gap::GapObj)
+      z = new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}()
+      z.parent = G
+      z.X = x_gap
+      return z
```

Comment:
If you swap the definitions of `elm` and `X` (making `elm` the last member of the struct), then I *think* you can change this to
```suggestion
      return new{elem_type(G.ring),dense_matrix_type(elem_type(G.ring))}(G, x_gap)
```

Reply:
Yes, but in this case I have to write a longer version for the other constructor (the one that defines `parent` and `elm`, but not `X`). At least one of them need to be long.